SEED-platform/seed
# feat(analyses): auto-poll analysis progress on frontend

#### Any background context you want to provide?

#### What's this PR do?
TLDR: for any page showing an analysis in a list or detailed view, SEED will now begin polling its status for updates so that it knows when to refresh it. This involved updating how progress data are created/tracked in the analysis pipelines.

SPECIFICALLY:
- add a method the the analyses_service which will continuously poll the status of an analysis until it determines theres no more progress data to track (`analyses_service.check_progress_loop(...)`)
- use method above in all views that render analyses so that they're refreshed when the status of the analysis changes
- part of accomplishing the above was adding a new endpoint `/analyses/<analysis id>/progress_key/` which returns the current `ProgressData` key (or null if there's no current `ProgressData` for the analysis).
  - This involved standardizing the progress data for all analysis pipelines. Progress data is now dependent on `Analysis.status`, and pipeline authors just need to use `pipeline.get_progress_data()` to get the current progress data.
  - As a result, pipeline authors must call the `AnalysisPipeline` base class methods for updating the status of the analysis instead of doing it directly (e.g. don't do `analysis.status = Analysis.COMPLETED; analysis.save()`). These methods are:
    - `set_analysis_status_to_ready`
    - `set_analysis_status_to_running`
    - `set_analysis_status_to_completed`
    - the existing `prepare_analysis`, `start_analysis`, `stop_analysis`, and `fail_analysis` methods already handle this for the remaining statuses.
   - the goal of these changes is to standardize the analysis pipeline web interface and hopefully simplify how pipelines are written - now pipeline authors just need to call these methods and not worry about finishing progress data, keeping track of progress data keys, updating the start/end time of the analysis, etc.

#### How should this be manually tested?
I'd recommend importing the anonymized JCC spreadsheet and meter data to start with. Then run try running different analyses and verify that the statuses are being updated correctly. It should NOT continue polling when the page is changed. It should NOT continue polling when the analysis finishes (fails, stopped, or completed).

Here are the pages affected (ie they should show live reloading as the analysis is being created/running/finishing):
- Analyses list page: `/app/#/accounts/<org id>/analyses`
- Analysis page: `/app/#/accounts/<org id>/analyses/<analysis id>`
- Analysis property view page: `/app/#/accounts/<org id>/analyses/<analysis id>/runs/<analysis property view id>` -- this one should actually refresh the whole page to show the results when the analysis is complete
- List of analyses for individual property: `/app/#/properties/<property_id>/analyses` -- this one should also refresh the whole page to view the results of the completed analysis
- Property details page with an analysis card: `/app/#/properties/<property_id>` (NOTE: might not see it here b/c the card is the most recently _completed_ analysis)

#### What are the relevant tickets?
#2816 

#### Screenshots (if appropriate)
![Screen Shot 2021-09-03 at 1 56 02 PM](https://user-images.githubusercontent.com/18518728/132059620-5da79510-ab14-4b07-a5e2-8437e65e0694.png)
![Screen Shot 2021-09-03 at 1 56 16 PM](https://user-images.githubusercontent.com/18518728/132059627-50499038-fe89-4308-8f82-62c1bf545a9a.png)
![Screen Shot 2021-09-03 at 1 56 54 PM](https://user-images.githubusercontent.com/18518728/132059631-c7e956b5-bc4c-41c7-ad16-27124b8c4d96.png)
![Screen Shot 2021-09-03 at 1 56 34 PM](https://user-images.githubusercontent.com/18518728/132059642-4c927495-f405-4d49-b5cf-091601ab1cfe.png)
